// @strapi/admin/dist/server/validation/transfer/token.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateTransferTokenUpdateInput = exports.validateTransferTokenCreationInput = void 0;
const utils_1 = require("@strapi/utils");
const constants_1 = __importDefault(require("../../services/constants"));
const transferTokenCreationSchema = utils_1.yup
    .object()
    .shape({
    name: utils_1.yup.string().min(1).required(),
    description: utils_1.yup.string().optional(),
    permissions: utils_1.yup
        .array()
        .min(1)
        .of(utils_1.yup.string().oneOf(Object.values(constants_1.default.TRANSFER_TOKEN_TYPE)))
        .required(),
    lifespan: utils_1.yup
        .number()
        .min(1)
        .oneOf(Object.values(constants_1.default.TRANSFER_TOKEN_LIFESPANS))
        .nullable(),
})
    .noUnknown()
    .strict();
const transferTokenUpdateSchema = utils_1.yup
    .object()
    .shape({
    name: utils_1.yup.string().min(1).notNull(),
    description: utils_1.yup.string().nullable(),
    permissions: utils_1.yup
        .array()
        .min(1)
        .of(utils_1.yup.string().oneOf(Object.values(constants_1.default.TRANSFER_TOKEN_TYPE)))
        .nullable(),
})
    .noUnknown()
    .strict();
exports.validateTransferTokenCreationInput = (0, utils_1.validateYupSchema)(transferTokenCreationSchema);
exports.validateTransferTokenUpdateInput = (0, utils_1.validateYupSchema)(transferTokenUpdateSchema);
exports.default = {
    validateTransferTokenCreationInput: exports.validateTransferTokenCreationInput,
    validateTransferTokenUpdateInput: exports.validateTransferTokenUpdateInput,
};
