// @strapi/strapi/dist/commands/builders/admin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = __importDefault(require("chalk"));
const admin_1 = __importDefault(require("@strapi/admin"));
const utils_1 = require("@strapi/utils");
const ee_1 = __importDefault(require("../../utils/ee"));
const addSlash_1 = __importDefault(require("../../utils/addSlash"));
const index_1 = __importDefault(require("../../index"));
const get_enabled_plugins_1 = require("../../core/loaders/plugins/get-enabled-plugins");
exports.default = async ({ buildDestDir, forceBuild = true, optimization, srcDir }) => {
    const strapiInstance = (0, index_1.default)({
        // Directories
        appDir: srcDir,
        distDir: buildDestDir,
        // Options
        autoReload: true,
        serveAdminPanel: false,
    });
    const plugins = await (0, get_enabled_plugins_1.getEnabledPlugins)(strapiInstance, { client: true });
    const env = strapiInstance.config.get('environment');
    const { serverUrl, adminPath } = (0, utils_1.getConfigUrls)(strapiInstance.config, true);
    console.log(`Building your admin UI with ${chalk_1.default.green(env)} configuration...`);
    // Always remove the .cache and build folders
    await admin_1.default.clean({ appDir: srcDir, buildDestDir });
    ee_1.default.init(srcDir);
    return admin_1.default
        .build({
        appDir: srcDir,
        buildDestDir,
        // front end build env is always production for now
        env: 'production',
        forceBuild,
        plugins,
        optimize: optimization,
        options: {
            backend: serverUrl,
            adminPath: (0, addSlash_1.default)(adminPath),
            telemetryIsDisabled: strapiInstance.telemetry.isDisabled,
        },
    })
        .then(() => {
        console.log('Admin UI built successfully');
    })
        .catch((err) => {
        console.error(err);
        process.exit(1);
    });
};
