// @strapi/admin/dist/server/services/permission.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.conditionProvider = exports.actionProvider = exports.engine = exports.sanitizePermission = exports.sectionsBuilder = exports.createPermissionsManager = exports.findUserPermissions = exports.findMany = exports.deleteByRolesIds = exports.deleteByIds = exports.createMany = exports.cleanPermissionsInDatabase = void 0;
const permission_1 = __importDefault(require("../domain/permission"));
const provider_1 = __importDefault(require("../domain/action/provider"));
const provider_2 = __importDefault(require("../domain/condition/provider"));
const permissions_manager_1 = __importDefault(require("./permission/permissions-manager"));
exports.createPermissionsManager = permissions_manager_1.default;
const engine_1 = __importDefault(require("./permission/engine"));
const sections_builder_1 = __importDefault(require("./permission/sections-builder"));
const queries_1 = require("./permission/queries");
Object.defineProperty(exports, "cleanPermissionsInDatabase", { enumerable: true, get: function () { return queries_1.cleanPermissionsInDatabase; } });
Object.defineProperty(exports, "createMany", { enumerable: true, get: function () { return queries_1.createMany; } });
Object.defineProperty(exports, "deleteByIds", { enumerable: true, get: function () { return queries_1.deleteByIds; } });
Object.defineProperty(exports, "deleteByRolesIds", { enumerable: true, get: function () { return queries_1.deleteByRolesIds; } });
Object.defineProperty(exports, "findMany", { enumerable: true, get: function () { return queries_1.findMany; } });
Object.defineProperty(exports, "findUserPermissions", { enumerable: true, get: function () { return queries_1.findUserPermissions; } });
const actionProvider = (0, provider_1.default)();
exports.actionProvider = actionProvider;
const conditionProvider = (0, provider_2.default)();
exports.conditionProvider = conditionProvider;
const sectionsBuilder = (0, sections_builder_1.default)();
exports.sectionsBuilder = sectionsBuilder;
const sanitizePermission = permission_1.default.sanitizePermissionFields;
exports.sanitizePermission = sanitizePermission;
const engine = (0, engine_1.default)({
    providers: { action: actionProvider, condition: conditionProvider },
});
exports.engine = engine;
