// @strapi/strapi/dist/commands/actions/ts/generate-types/action.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const typescript_utils_1 = __importDefault(require("@strapi/typescript-utils"));
const index_1 = __importDefault(require("../../../../index"));
exports.default = async ({ debug, silent, verbose, outDir }) => {
    if ((debug || verbose) && silent) {
        console.error('Flags conflict: both silent and debug mode are enabled, exiting...');
        process.exit(1);
    }
    const appContext = await index_1.default.compile({ ignoreDiagnostics: true });
    const app = await (0, index_1.default)(appContext).register();
    await typescript_utils_1.default.generators.generate({
        strapi: app,
        pwd: appContext.appDir,
        rootDir: outDir ?? undefined,
        logger: {
            silent,
            // TODO V5: verbose is deprecated and should be removed
            debug: debug || verbose,
        },
        artifacts: { contentTypes: true, components: true },
    });
    await app.destroy();
};
