// @strapi/admin/dist/server/validation/authentication/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateRenewTokenInput = exports.validateResetPasswordInput = exports.validateForgotPasswordInput = exports.validateRegistrationInfoQuery = exports.validateAdminRegistrationInput = exports.validateRegistrationInput = void 0;
const register_1 = require("./register");
Object.defineProperty(exports, "validateRegistrationInput", { enumerable: true, get: function () { return register_1.validateRegistrationInput; } });
Object.defineProperty(exports, "validateAdminRegistrationInput", { enumerable: true, get: function () { return register_1.validateAdminRegistrationInput; } });
Object.defineProperty(exports, "validateRegistrationInfoQuery", { enumerable: true, get: function () { return register_1.validateRegistrationInfoQuery; } });
var forgot_password_1 = require("./forgot-password");
Object.defineProperty(exports, "validateForgotPasswordInput", { enumerable: true, get: function () { return __importDefault(forgot_password_1).default; } });
var reset_password_1 = require("./reset-password");
Object.defineProperty(exports, "validateResetPasswordInput", { enumerable: true, get: function () { return __importDefault(reset_password_1).default; } });
var renew_token_1 = require("./renew-token");
Object.defineProperty(exports, "validateRenewTokenInput", { enumerable: true, get: function () { return __importDefault(renew_token_1).default; } });
