// @strapi/admin/dist/server/validation/authentication/register.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateAdminRegistrationInput = exports.validateRegistrationInfoQuery = exports.validateRegistrationInput = void 0;
const utils_1 = require("@strapi/utils");
const common_validators_1 = __importDefault(require("../common-validators"));
const registrationSchema = utils_1.yup
    .object()
    .shape({
    registrationToken: utils_1.yup.string().required(),
    userInfo: utils_1.yup
        .object()
        .shape({
        firstname: common_validators_1.default.firstname.required(),
        lastname: common_validators_1.default.lastname.nullable(),
        password: common_validators_1.default.password.required(),
    })
        .required()
        .noUnknown(),
})
    .noUnknown();
const registrationInfoQuerySchema = utils_1.yup
    .object()
    .shape({
    registrationToken: utils_1.yup.string().required(),
})
    .required()
    .noUnknown();
const adminRegistrationSchema = utils_1.yup
    .object()
    .shape({
    email: common_validators_1.default.email.required(),
    firstname: common_validators_1.default.firstname.required(),
    lastname: common_validators_1.default.lastname.nullable(),
    password: common_validators_1.default.password.required(),
})
    .required()
    .noUnknown();
exports.validateRegistrationInput = (0, utils_1.validateYupSchema)(registrationSchema);
exports.validateRegistrationInfoQuery = (0, utils_1.validateYupSchema)(registrationInfoQuerySchema);
exports.validateAdminRegistrationInput = (0, utils_1.validateYupSchema)(adminRegistrationSchema);
exports.default = {
    validateRegistrationInput: exports.validateRegistrationInput,
    validateRegistrationInfoQuery: exports.validateRegistrationInfoQuery,
    validateAdminRegistrationInput: exports.validateAdminRegistrationInput,
};
