// @strapi/strapi/dist/services/server/koa.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fp_1 = require("lodash/fp");
const koa_1 = __importDefault(require("koa"));
const http_errors_1 = __importDefault(require("http-errors"));
const delegates_1 = __importDefault(require("delegates"));
const statuses_1 = __importDefault(require("statuses"));
const errors_1 = require("../errors");
const addCustomMethods = (app) => {
    const delegator = (0, delegates_1.default)(app.context, 'response');
    /* errors */
    statuses_1.default.codes
        .filter((code) => code >= 400 && code < 600)
        .forEach((code) => {
        const name = (0, statuses_1.default)(code);
        const camelCasedName = (0, fp_1.camelCase)(name);
        app.response[camelCasedName] = function responseCode(message, details = {}) {
            const httpError = (0, http_errors_1.default)(code, message, { details });
            const { status, body } = (0, errors_1.formatHttpError)(httpError);
            this.status = status;
            this.body = body;
        };
        delegator.method(camelCasedName);
    });
    /* send, created, deleted */
    app.response.send = function send(data, status = 200) {
        this.status = status;
        this.body = data;
    };
    app.response.created = function created(data) {
        this.status = 201;
        this.body = data;
    };
    app.response.deleted = function deleted(data) {
        if ((0, fp_1.isNil)(data)) {
            this.status = 204;
        }
        else {
            this.status = 200;
            this.body = data;
        }
    };
    delegator.method('send').method('created').method('deleted');
    return app;
};
const createKoaApp = ({ proxy, keys }) => {
    const app = new koa_1.default({ proxy });
    app.keys = keys;
    addCustomMethods(app);
    return app;
};
exports.default = createKoaApp;
