// @strapi/admin/dist/server/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.middlewares = exports.contentTypes = exports.controllers = exports.services = exports.routes = exports.policies = exports.config = exports.destroy = exports.register = exports.bootstrap = void 0;
var bootstrap_1 = require("./bootstrap");
Object.defineProperty(exports, "bootstrap", { enumerable: true, get: function () { return __importDefault(bootstrap_1).default; } });
var register_1 = require("./register");
Object.defineProperty(exports, "register", { enumerable: true, get: function () { return __importDefault(register_1).default; } });
var destroy_1 = require("./destroy");
Object.defineProperty(exports, "destroy", { enumerable: true, get: function () { return __importDefault(destroy_1).default; } });
var config_1 = require("./config");
Object.defineProperty(exports, "config", { enumerable: true, get: function () { return __importDefault(config_1).default; } });
var policies_1 = require("./policies");
Object.defineProperty(exports, "policies", { enumerable: true, get: function () { return __importDefault(policies_1).default; } });
var routes_1 = require("./routes");
Object.defineProperty(exports, "routes", { enumerable: true, get: function () { return __importDefault(routes_1).default; } });
var services_1 = require("./services");
Object.defineProperty(exports, "services", { enumerable: true, get: function () { return __importDefault(services_1).default; } });
var controllers_1 = require("./controllers");
Object.defineProperty(exports, "controllers", { enumerable: true, get: function () { return __importDefault(controllers_1).default; } });
var content_types_1 = require("./content-types");
Object.defineProperty(exports, "contentTypes", { enumerable: true, get: function () { return __importDefault(content_types_1).default; } });
var middlewares_1 = require("./middlewares");
Object.defineProperty(exports, "middlewares", { enumerable: true, get: function () { return __importDefault(middlewares_1).default; } });
