// @strapi/strapi/dist/utils/get-dirs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const getDirs = ({ app: appDir, dist: distDir }, { strapi }) => ({
    dist: {
        root: distDir,
        src: (0, path_1.join)(distDir, 'src'),
        api: (0, path_1.join)(distDir, 'src', 'api'),
        components: (0, path_1.join)(distDir, 'src', 'components'),
        extensions: (0, path_1.join)(distDir, 'src', 'extensions'),
        policies: (0, path_1.join)(distDir, 'src', 'policies'),
        middlewares: (0, path_1.join)(distDir, 'src', 'middlewares'),
        config: (0, path_1.join)(distDir, 'config'),
    },
    app: {
        root: appDir,
        src: (0, path_1.join)(appDir, 'src'),
        api: (0, path_1.join)(appDir, 'src', 'api'),
        components: (0, path_1.join)(appDir, 'src', 'components'),
        extensions: (0, path_1.join)(appDir, 'src', 'extensions'),
        policies: (0, path_1.join)(appDir, 'src', 'policies'),
        middlewares: (0, path_1.join)(appDir, 'src', 'middlewares'),
        config: (0, path_1.join)(appDir, 'config'),
    },
    static: {
        public: (0, path_1.resolve)(appDir, strapi.config.get('server.dirs.public')),
    },
});
exports.default = getDirs;
