// @strapi/admin/dist/server/domain/condition/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = exports.assignConditionId = exports.computeConditionId = exports.sanitizeConditionAttributes = exports.conditionFields = exports.getDefaultConditionAttributes = void 0;
const fp_1 = require("lodash/fp");
const DEFAULT_CATEGORY = 'default';
/**
 * Get the default value used for every condition
 * @return {Condition}
 */
const getDefaultConditionAttributes = () => ({
    category: DEFAULT_CATEGORY,
});
exports.getDefaultConditionAttributes = getDefaultConditionAttributes;
/**
 * Get the list of all the valid attributes of a {@link Condition}
 * @return {string[]}
 */
exports.conditionFields = ['id', 'displayName', 'handler', 'plugin', 'category'];
/**
 * Remove unwanted attributes from a {@link Condition}
 */
exports.sanitizeConditionAttributes = (0, fp_1.pick)(exports.conditionFields);
const computeConditionId = (condition) => {
    const { name, plugin } = condition;
    if (!plugin) {
        return `api::${name}`;
    }
    if (plugin === 'admin') {
        return `admin::${name}`;
    }
    return `plugin::${plugin}.${name}`;
};
exports.computeConditionId = computeConditionId;
/**
 * Assign an id attribute to a {@link CreateConditionPayload} object
 * @param  attrs - Payload used to create a condition
 */
const assignConditionId = (attrs) => {
    const condition = (0, fp_1.set)('id', (0, exports.computeConditionId)(attrs), attrs);
    return condition;
};
exports.assignConditionId = assignConditionId;
/**
 * Transform the given attributes into a domain representation of a Condition
 * @param payload - The condition payload containing the attributes needed to create a {@link Condition}
 */
exports.create = (0, fp_1.pipe)(exports.assignConditionId, exports.sanitizeConditionAttributes, (0, fp_1.merge)((0, exports.getDefaultConditionAttributes)()));
exports.default = {
    assignConditionId: exports.assignConditionId,
    computeConditionId: exports.computeConditionId,
    conditionFields: exports.conditionFields,
    create: exports.create,
    getDefaultConditionAttributes: exports.getDefaultConditionAttributes,
    sanitizeConditionAttributes: exports.sanitizeConditionAttributes,
};
