// @strapi/strapi/dist/commands/utils/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLogger = void 0;
const chalk_1 = __importDefault(require("chalk"));
const createLogger = (options = {}) => {
    const { silent = false, debug = false, timestamp = true } = options;
    const state = { errors: 0, warning: 0 };
    return {
        get warnings() {
            return state.warning;
        },
        get errors() {
            return state.errors;
        },
        debug(...args) {
            if (silent || !debug) {
                return;
            }
            console.log(chalk_1.default.cyan(`[DEBUG]${timestamp ? `\t[${new Date().toISOString()}]` : ''}`), ...args);
        },
        info(...args) {
            if (silent) {
                return;
            }
            console.info(chalk_1.default.blue(`[INFO]${timestamp ? `\t[${new Date().toISOString()}]` : ''}`), ...args);
        },
        log(...args) {
            if (silent) {
                return;
            }
            console.info(chalk_1.default.blue(`${timestamp ? `\t[${new Date().toISOString()}]` : ''}`), ...args);
        },
        warn(...args) {
            state.warning += 1;
            if (silent) {
                return;
            }
            console.warn(chalk_1.default.yellow(`[WARN]${timestamp ? `\t[${new Date().toISOString()}]` : ''}`), ...args);
        },
        error(...args) {
            state.errors += 1;
            if (silent) {
                return;
            }
            console.error(chalk_1.default.red(`[ERROR]${timestamp ? `\t[${new Date().toISOString()}]` : ''}`), ...args);
        },
    };
};
exports.createLogger = createLogger;
