// @strapi/data-transfer/dist/commands/transfer/action.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const fp_1 = require("lodash/fp");
const engineDatatransfer = __importStar(require("../../engine"));
const strapiDatatransfer = __importStar(require("../../strapi"));
const data_transfer_1 = require("../data-transfer");
const helpers_1 = require("../helpers");
const { createTransferEngine } = engineDatatransfer;
const { providers: { createRemoteStrapiDestinationProvider, createLocalStrapiSourceProvider, createLocalStrapiDestinationProvider, createRemoteStrapiSourceProvider, }, } = strapiDatatransfer;
/**
 * Transfer command.
 *
 * Transfers data between local Strapi and remote Strapi instances
 */
exports.default = async (opts) => {
    // Validate inputs from Commander
    if (!(0, fp_1.isObject)(opts)) {
        (0, helpers_1.exitWith)(1, 'Could not parse command arguments');
    }
    if (!(opts.from || opts.to) || (opts.from && opts.to)) {
        (0, helpers_1.exitWith)(1, 'Exactly one source (from) or destination (to) option must be provided');
    }
    const strapi = await (0, data_transfer_1.createStrapiInstance)();
    let source;
    let destination;
    // if no URL provided, use local Strapi
    if (!opts.from) {
        source = createLocalStrapiSourceProvider({
            getStrapi: () => strapi,
        });
    }
    // if URL provided, set up a remote source provider
    else {
        if (!opts.fromToken) {
            (0, helpers_1.exitWith)(1, 'Missing token for remote destination');
        }
        source = createRemoteStrapiSourceProvider({
            getStrapi: () => strapi,
            url: opts.from,
            auth: {
                type: 'token',
                token: opts.fromToken,
            },
        });
    }
    // if no URL provided, use local Strapi
    if (!opts.to) {
        destination = createLocalStrapiDestinationProvider({
            getStrapi: () => strapi,
            strategy: 'restore',
            restore: (0, data_transfer_1.parseRestoreFromOptions)(opts),
        });
    }
    // if URL provided, set up a remote destination provider
    else {
        if (!opts.toToken) {
            (0, helpers_1.exitWith)(1, 'Missing token for remote destination');
        }
        destination = createRemoteStrapiDestinationProvider({
            url: opts.to,
            auth: {
                type: 'token',
                token: opts.toToken,
            },
            strategy: 'restore',
            restore: (0, data_transfer_1.parseRestoreFromOptions)(opts),
        });
    }
    if (!source || !destination) {
        (0, helpers_1.exitWith)(1, 'Could not create providers');
    }
    const engine = createTransferEngine(source, destination, {
        versionStrategy: 'exact',
        schemaStrategy: 'strict',
        exclude: opts.exclude,
        only: opts.only,
        throttle: opts.throttle,
        transforms: {
            links: [
                {
                    filter(link) {
                        return (!data_transfer_1.DEFAULT_IGNORED_CONTENT_TYPES.includes(link.left.type) &&
                            !data_transfer_1.DEFAULT_IGNORED_CONTENT_TYPES.includes(link.right.type));
                    },
                },
            ],
            entities: [
                {
                    filter(entity) {
                        return !data_transfer_1.DEFAULT_IGNORED_CONTENT_TYPES.includes(entity.type);
                    },
                },
            ],
        },
    });
    engine.diagnostics.onDiagnostic((0, data_transfer_1.formatDiagnostic)('transfer'));
    const progress = engine.progress.stream;
    const { updateLoader } = (0, data_transfer_1.loadersFactory)();
    engine.onSchemaDiff((0, data_transfer_1.getDiffHandler)(engine, { force: opts.force, action: 'transfer' }));
    engine.addErrorHandler('ASSETS_DIRECTORY_ERR', (0, data_transfer_1.getAssetsBackupHandler)(engine, { force: opts.force, action: 'transfer' }));
    progress.on(`stage::start`, ({ stage, data }) => {
        updateLoader(stage, data).start();
    });
    progress.on('stage::finish', ({ stage, data }) => {
        updateLoader(stage, data).succeed();
    });
    progress.on('stage::progress', ({ stage, data }) => {
        updateLoader(stage, data);
    });
    progress.on('stage::error', ({ stage, data }) => {
        updateLoader(stage, data).fail();
    });
    progress.on('transfer::start', async () => {
        console.log(`Starting transfer...`);
        await strapi.telemetry.send('didDEITSProcessStart', (0, data_transfer_1.getTransferTelemetryPayload)(engine));
    });
    let results;
    try {
        // Abort transfer if user interrupts process
        (0, data_transfer_1.setSignalHandler)(() => (0, data_transfer_1.abortTransfer)({ engine, strapi }));
        results = await engine.transfer();
        // Note: we need to await telemetry or else the process ends before it is sent
        await strapi.telemetry.send('didDEITSProcessFinish', (0, data_transfer_1.getTransferTelemetryPayload)(engine));
        try {
            const table = (0, data_transfer_1.buildTransferTable)(results.engine);
            console.log(table?.toString());
        }
        catch (e) {
            console.error('There was an error displaying the results of the transfer.');
        }
        (0, helpers_1.exitWith)(0, (0, data_transfer_1.exitMessageText)('transfer'));
    }
    catch (e) {
        await strapi.telemetry.send('didDEITSProcessFail', (0, data_transfer_1.getTransferTelemetryPayload)(engine));
        (0, helpers_1.exitWith)(1, (0, data_transfer_1.exitMessageText)('transfer', true));
    }
};
