// @strapi/admin/dist/server/controllers/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const admin = __importStar(require("./admin"));
const api_token_1 = __importDefault(require("./api-token"));
const authenticated_user_1 = __importDefault(require("./authenticated-user"));
const authentication_1 = __importDefault(require("./authentication"));
const permission_1 = __importDefault(require("./permission"));
const role_1 = __importDefault(require("./role"));
const transfer_1 = __importDefault(require("./transfer"));
const user_1 = __importDefault(require("./user"));
const webhooks_1 = __importDefault(require("./webhooks"));
const content_api_1 = __importDefault(require("./content-api"));
exports.default = {
    admin,
    'api-token': api_token_1.default,
    'authenticated-user': authenticated_user_1.default,
    authentication: authentication_1.default,
    permission: permission_1.default,
    role: role_1.default,
    transfer: transfer_1.default,
    user: user_1.default,
    webhooks: webhooks_1.default,
    'content-api': content_api_1.default,
};
