// @strapi/strapi/dist/services/worker-queue.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Simple worker queue in memory
 */
const debug_1 = __importDefault(require("debug"));
const debug = (0, debug_1.default)('strapi:worker-queue');
const noop = () => { };
class WorkerQueue {
    logger;
    worker;
    concurrency;
    running;
    queue;
    constructor({ logger, concurrency = 5 }) {
        debug('Initialize worker queue');
        this.logger = logger;
        this.worker = noop;
        this.concurrency = concurrency;
        this.running = 0;
        this.queue = [];
    }
    subscribe(worker) {
        debug('Subscribe to worker queue');
        this.worker = worker;
    }
    enqueue(payload) {
        debug('Enqueue event in worker queue');
        if (this.running < this.concurrency) {
            this.running += 1;
            this.execute(payload);
        }
        else {
            this.queue.unshift(payload);
        }
    }
    pop() {
        debug('Pop worker queue and execute');
        const payload = this.queue.pop();
        if (payload) {
            this.execute(payload);
        }
        else {
            this.running -= 1;
        }
    }
    async execute(payload) {
        debug('Execute worker');
        try {
            await this.worker(payload);
        }
        catch (error) {
            this.logger.error(error);
        }
        finally {
            this.pop();
        }
    }
}
exports.default = WorkerQueue;
