// @strapi/strapi/dist/middlewares/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.middlewares = void 0;
const compression_1 = require("./compression");
const cors_1 = require("./cors");
const errors_1 = require("./errors");
const favicon_1 = require("./favicon");
const ip_1 = require("./ip");
const logger_1 = require("./logger");
const powered_by_1 = require("./powered-by");
const body_1 = require("./body");
const query_1 = require("./query");
const response_time_1 = require("./response-time");
const responses_1 = require("./responses");
const security_1 = require("./security");
const session_1 = require("./session");
const public_1 = require("./public");
exports.middlewares = {
    compression: compression_1.compression,
    cors: cors_1.cors,
    errors: errors_1.errors,
    favicon: favicon_1.favicon,
    ip: ip_1.ip,
    logger: logger_1.logger,
    poweredBy: powered_by_1.poweredBy,
    body: body_1.body,
    query: query_1.query,
    responseTime: response_time_1.responseTime,
    responses: responses_1.responses,
    security: security_1.security,
    session: session_1.session,
    public: public_1.publicStatic,
};
