// @strapi/strapi/dist/middlewares/session.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.session = void 0;
const fp_1 = require("lodash/fp");
const koa_session_1 = __importDefault(require("koa-session"));
const defaultConfig = {
    key: 'koa.sess',
    maxAge: 86400000,
    autoCommit: true,
    overwrite: true,
    httpOnly: true,
    signed: true,
    rolling: false,
    renew: false,
    secure: process.env.NODE_ENV === 'production',
    sameSite: undefined,
};
const session = (userConfig, { strapi }) => {
    const { keys } = strapi.server.app;
    if (!(0, fp_1.isArray)(keys) || (0, fp_1.isEmpty)(keys) || keys.some(fp_1.isEmpty)) {
        throw new Error(`App keys are required. Please set app.keys in config/server.js (ex: keys: ['myKeyA', 'myKeyB'])`);
    }
    const config = { ...defaultConfig, ...userConfig };
    strapi.server.use((0, koa_session_1.default)(config, strapi.server.app));
};
exports.session = session;
