// @strapi/admin/dist/server/services/permission/sections-builder/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toSubjectTemplate = exports.getValidOptions = exports.hasProperty = exports.isNotInSubjects = exports.resolveContentType = exports.isOfKind = void 0;
const fp_1 = require("lodash/fp");
const isOfKind = (kind) => (0, fp_1.matchesProperty)('kind', kind);
exports.isOfKind = isOfKind;
const resolveContentType = (uid) => strapi.contentTypes[uid];
exports.resolveContentType = resolveContentType;
const isNotInSubjects = (subjects) => (uid) => !subjects.find((subject) => subject.uid === uid);
exports.isNotInSubjects = isNotInSubjects;
const hasProperty = (0, fp_1.curry)((property, subject) => {
    return !!subject.properties.find((prop) => prop.value === property);
});
exports.hasProperty = hasProperty;
const getValidOptions = (0, fp_1.pick)(['applyToProperties']);
exports.getValidOptions = getValidOptions;
const toSubjectTemplate = (ct) => ({
    uid: ct.uid,
    label: ct.info.singularName,
    properties: [],
});
exports.toSubjectTemplate = toSubjectTemplate;
