// @strapi/admin/dist/server/validation/api-tokens.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateApiTokenUpdateInput = exports.validateApiTokenCreationInput = void 0;
const utils_1 = require("@strapi/utils");
const constants_1 = __importDefault(require("../services/constants"));
const apiTokenCreationSchema = utils_1.yup
    .object()
    .shape({
    name: utils_1.yup.string().min(1).required(),
    description: utils_1.yup.string().optional(),
    type: utils_1.yup.string().oneOf(Object.values(constants_1.default.API_TOKEN_TYPE)).required(),
    permissions: utils_1.yup.array().of(utils_1.yup.string()).nullable(),
    lifespan: utils_1.yup.number().min(1).oneOf(Object.values(constants_1.default.API_TOKEN_LIFESPANS)).nullable(),
})
    .noUnknown()
    .strict();
const apiTokenUpdateSchema = utils_1.yup
    .object()
    .shape({
    name: utils_1.yup.string().min(1).notNull(),
    description: utils_1.yup.string().nullable(),
    type: utils_1.yup.string().oneOf(Object.values(constants_1.default.API_TOKEN_TYPE)).notNull(),
    permissions: utils_1.yup.array().of(utils_1.yup.string()).nullable(),
})
    .noUnknown()
    .strict();
exports.validateApiTokenCreationInput = (0, utils_1.validateYupSchema)(apiTokenCreationSchema);
exports.validateApiTokenUpdateInput = (0, utils_1.validateYupSchema)(apiTokenUpdateSchema);
exports.default = {
    validateApiTokenCreationInput: exports.validateApiTokenCreationInput,
    validateApiTokenUpdateInput: exports.validateApiTokenUpdateInput,
};
