// @strapi/admin/dist/server/validation/project-settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateUpdateProjectSettingsImagesDimensions = exports.validateUpdateProjectSettingsFiles = exports.validateUpdateProjectSettings = void 0;
const utils_1 = require("@strapi/utils");
const MAX_IMAGE_WIDTH = 750;
const MAX_IMAGE_HEIGHT = MAX_IMAGE_WIDTH;
const MAX_IMAGE_FILE_SIZE = 1024 * 1024; // 1Mo
const ALLOWED_IMAGE_FILE_TYPES = ['image/jpeg', 'image/png', 'image/svg+xml'];
const updateProjectSettings = utils_1.yup
    .object({
    menuLogo: utils_1.yup.string(),
    authLogo: utils_1.yup.string(),
})
    .noUnknown();
const updateProjectSettingsLogo = utils_1.yup.object({
    name: utils_1.yup.string(),
    type: utils_1.yup.string().oneOf(ALLOWED_IMAGE_FILE_TYPES),
    size: utils_1.yup.number().max(MAX_IMAGE_FILE_SIZE),
});
const updateProjectSettingsFiles = utils_1.yup
    .object({
    menuLogo: updateProjectSettingsLogo,
    authLogo: updateProjectSettingsLogo,
})
    .noUnknown();
const logoDimensions = utils_1.yup.object({
    width: utils_1.yup.number().max(MAX_IMAGE_WIDTH),
    height: utils_1.yup.number().max(MAX_IMAGE_HEIGHT),
});
const updateProjectSettingsImagesDimensions = utils_1.yup
    .object({
    menuLogo: logoDimensions,
    authLogo: logoDimensions,
})
    .noUnknown();
exports.validateUpdateProjectSettings = (0, utils_1.validateYupSchemaSync)(updateProjectSettings);
exports.validateUpdateProjectSettingsFiles = (0, utils_1.validateYupSchemaSync)(updateProjectSettingsFiles);
exports.validateUpdateProjectSettingsImagesDimensions = (0, utils_1.validateYupSchemaSync)(updateProjectSettingsImagesDimensions);
exports.default = {
    validateUpdateProjectSettings: exports.validateUpdateProjectSettings,
    validateUpdateProjectSettingsFiles: exports.validateUpdateProjectSettingsFiles,
    validateUpdateProjectSettingsImagesDimensions: exports.validateUpdateProjectSettingsImagesDimensions,
};
