// @strapi/admin/dist/server/validation/role.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateRoleDeleteInput = exports.validateRolesDeleteInput = exports.validateRoleUpdateInput = exports.validateRoleCreateInput = void 0;
const utils_1 = require("@strapi/utils");
const roleCreateSchema = utils_1.yup
    .object()
    .shape({
    name: utils_1.yup.string().min(1).required(),
    description: utils_1.yup.string().nullable(),
})
    .noUnknown();
const rolesDeleteSchema = utils_1.yup
    .object()
    .shape({
    ids: utils_1.yup
        .array()
        .of(utils_1.yup.strapiID())
        .min(1)
        .required()
        .test('roles-deletion-checks', 'Roles deletion checks have failed', async function (ids) {
        try {
            await strapi.admin.services.role.checkRolesIdForDeletion(ids);
        }
        catch (e) {
            // @ts-expect-error
            return this.createError({ path: 'ids', message: e.message });
        }
        return true;
    }),
})
    .noUnknown();
const roleDeleteSchema = utils_1.yup
    .strapiID()
    .required()
    .test('no-admin-single-delete', 'Role deletion checks have failed', async function (id) {
    try {
        await strapi.admin.services.role.checkRolesIdForDeletion([id]);
    }
    catch (e) {
        // @ts-expect-error
        return this.createError({ path: 'id', message: e.message });
    }
    return true;
});
const roleUpdateSchema = utils_1.yup
    .object()
    .shape({
    name: utils_1.yup.string().min(1),
    description: utils_1.yup.string().nullable(),
})
    .noUnknown();
exports.validateRoleCreateInput = (0, utils_1.validateYupSchema)(roleCreateSchema);
exports.validateRoleUpdateInput = (0, utils_1.validateYupSchema)(roleUpdateSchema);
exports.validateRolesDeleteInput = (0, utils_1.validateYupSchema)(rolesDeleteSchema);
exports.validateRoleDeleteInput = (0, utils_1.validateYupSchema)(roleDeleteSchema);
exports.default = {
    validateRoleUpdateInput: exports.validateRoleUpdateInput,
    validateRoleCreateInput: exports.validateRoleCreateInput,
    validateRolesDeleteInput: exports.validateRolesDeleteInput,
    validateRoleDeleteInput: exports.validateRoleDeleteInput,
};
