// @strapi/admin/dist/server/validation/user.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.schemas = exports.validateUsersDeleteInput = exports.validateUserUpdateInput = exports.validateProfileUpdateInput = exports.validateUserCreationInput = void 0;
const fp_1 = require("lodash/fp");
const utils_1 = require("@strapi/utils");
const common_validators_1 = __importDefault(require("./common-validators"));
const userCreationSchema = utils_1.yup
    .object()
    .shape({
    email: common_validators_1.default.email.required(),
    firstname: common_validators_1.default.firstname.required(),
    lastname: common_validators_1.default.lastname,
    roles: common_validators_1.default.roles.min(1),
    preferedLanguage: utils_1.yup.string().nullable(),
})
    .noUnknown();
const profileUpdateSchema = utils_1.yup
    .object()
    .shape({
    email: common_validators_1.default.email.notNull(),
    firstname: common_validators_1.default.firstname.notNull(),
    lastname: common_validators_1.default.lastname.nullable(),
    username: common_validators_1.default.username.nullable(),
    password: common_validators_1.default.password.notNull(),
    currentPassword: utils_1.yup
        .string()
        .when('password', (password, schema) => !(0, fp_1.isUndefined)(password) ? schema.required() : schema)
        .notNull(),
    preferedLanguage: utils_1.yup.string().nullable(),
})
    .noUnknown();
const userUpdateSchema = utils_1.yup
    .object()
    .shape({
    email: common_validators_1.default.email.notNull(),
    firstname: common_validators_1.default.firstname.notNull(),
    lastname: common_validators_1.default.lastname.nullable(),
    username: common_validators_1.default.username.nullable(),
    password: common_validators_1.default.password.notNull(),
    isActive: utils_1.yup.bool().notNull(),
    roles: common_validators_1.default.roles.min(1).notNull(),
})
    .noUnknown();
const usersDeleteSchema = utils_1.yup
    .object()
    .shape({
    ids: utils_1.yup.array().of(utils_1.yup.strapiID()).min(1).required(),
})
    .noUnknown();
exports.validateUserCreationInput = (0, utils_1.validateYupSchema)(userCreationSchema);
exports.validateProfileUpdateInput = (0, utils_1.validateYupSchema)(profileUpdateSchema);
exports.validateUserUpdateInput = (0, utils_1.validateYupSchema)(userUpdateSchema);
exports.validateUsersDeleteInput = (0, utils_1.validateYupSchema)(usersDeleteSchema);
exports.schemas = {
    userCreationSchema,
    usersDeleteSchema,
    userUpdateSchema,
};
exports.default = {
    validateUserCreationInput: exports.validateUserCreationInput,
    validateProfileUpdateInput: exports.validateProfileUpdateInput,
    validateUserUpdateInput: exports.validateUserUpdateInput,
    validateUsersDeleteInput: exports.validateUsersDeleteInput,
    schemas: exports.schemas,
};
